README
$Id: README,v 1.2 2005/06/23 07:07:18 kees-guest Exp $

METAR reports are meteorogical weather reports for aviation. Metar is a small
program which downloads weather reports for user-specified stations and
optionally decodes them into a human-readable format.

Currently, metar supports decoding date/time, wind, visibility, cloud layers,
temperature, air pressure and weather phenomena, such as rain, fog, etc. Also,
more work in the area of clouds need to be done, as support for Cumulus or
Cumulunimbus is not yet decoded.

There is a distinct difference between European METAR reports and US Metar
reports. A US METAR report typically looks like:

KJFK 141751Z 33017G22KT 10SM FEW050 02/M10 A3005 RMK AO2 SLP174 T00221100
10033 20017 51007

This METAR can be decoded as follows:

KJFK	    John F. Kennedy Airport, New York, New York, This is the ICAO
            identifier of the airport.

141751Z     The report is of the 14th of the month, at 17:51Z (which is 12:51pm
            local time in New York in the winter.)

33017G22KT  The wind is coming from the direction of 330 degrees (Northwest)
            at 17 knots, with wind gusts to 22 knots.

10SM        Visibility at the ground is clear (10 statue miles or more)

FEW050      FEW clouds at 5000 feet. Few is METAR-speak for 2/8 cloud cover.

02/M10      The temperature is 2 degrees Celcius, with the dewpoint at -10C.

A3005       The altimeter setting (air pressure) is 30.05 inches Hg

RMK etc     Remarks local to the field.



A European METAR report:

EHAM 141825Z 18010KT 2700 BR SCT003 BKN004 04/04 Q1024 BECMG BKN003

EHAM	 	Amsterdam Schiphol Airport, The Netherlands.

141825Z	    The report is of the 14th of the month, at 18:25Z (which is 19:25
            local time in Amsterdam in the winter.

18010KT	    The wind is coming from 180 degrees (due South) at 10 knots.

2700        Visibility at ground level is 2700 meters.

BR          Mist (visibility at least 5/8 statue mile)

SCT003      Scattered clouds at 300 ft (SCT is 4/8 cover)

BKN004      Broken clouds at 400 ft (BKN is 5/8 through 7/8 cover)

04/04	    Temperature 4 degrees C, dewpoint 4 C

Q1024	    Air pressure at 1024 hPa

BECMG BKN003 Becoming BK003, cloud layer is changing to BKN at 300 ft.


The program currently supports the following command line paramters:

-d 	Decode METARs
-h  Show a short usage summary
-v  Be verbose


Currently recognized weather phenomena are:

	Descriptors
		MI	Shallow
		BL	Blowing
		BC	Patches
		SH	Showers
		PR	Partial
		DR	Drifting
		TS	Thunderstorm
		FZ	Freezing
		VC Vicinity

	Weather phenomena
		DZ	Drizzle
		IC	Ice Crystals
		UP	Unkniwn
		RA	Rain
		PL	Ice Pellets
		SN	Snow
		GR	Hail
		SG	Snow Grains
		GS	Small hail/snow pellets
	
	Obscuration
		BR	Mist (or vis <= 5/8 sm)
		SA	Sand
		FU	Smoke
		HZ	Haze
		VA	Volcanic Ash
		PY	Spray
		DU	Widespread Dust

	Other
		SQ	Squall
		FC	Funnel Cloud
		SS	Sand storm
		+FC	Tornado/Watersprout
		DS	Dust storm
		PO	Well developed dust/sand swirls

Metar sources and bug tracking are gracefully hosted by the Debian project.
Please visit http://alioth.debian.org/projects/metar/ for more information.

% END-OF-FILE


NEWS
$Id: NEWS,v 1.2 2005/04/08 07:19:21 kees-guest Exp $

08-apr-2005   After receiving a few bug reports and some wishlist items via
              the Debian bug tracking system, I decided to prepare a new
			  version of the package. Most important changes:

			  - metar now defaults to HTTP instead of FTP
			  - download URL is settable via an environment variable
			  - CAVOK and variables winds are now properly decoded

14-dec-2004	  First version of the metar package.


src/main.c
/* metar.c -- metar decoder
   $Id: main.c,v 1.9 2006/04/05 20:30:28 kees-guest Exp $
   Copyright 2004,2005 Kees Leune <kees@leune.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#include <stdio.h>
#include <ctype.h>
#include <curl/curl.h>
#include <sys/types.h>
#include <regex.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "metar.h"

/* global variable so we dont have to mess with parameter passing */
char noaabuffer[METAR_MAXSIZE];

/* command line args */
int  decode=0;
int  verbose=0;

char *strupc(char *line) {
   char *p;
   for (p=line; *p; p++) *p=toupper(*p);
   return line;
}

/* show brief usage info */
void usage(char *name) {
	printf("$Id: main.c,v 1.9 2006/04/05 20:30:28 kees-guest Exp $\n");
	printf("Usage: %s options\n", name);
	printf("Options\n");
	printf("   -d        decode metar\n");
	printf("   -h        show this help\n");
	printf("   -v        be verbose\n");
	printf("Example: %s -d ehgr\n", name);
}


/* place NOAA data in buffer */
int receiveData(void *buffer, size_t size, size_t nmemb, void *stream) {
	size *= nmemb;
	size = (size <= METAR_MAXSIZE) ? size : METAR_MAXSIZE;
	strncpy(noaabuffer, buffer, size);
	return size;
}


/* fetch NOAA report */
int download_Metar(char *station) {
    CURL *curlhandle = NULL;
	CURLcode res;
    char url[URL_MAXSIZE];
	char tmp[URL_MAXSIZE];

    curlhandle = curl_easy_init();
	if (!curlhandle) return 1;

	memset(tmp, 0x0, URL_MAXSIZE);
	if (getenv("METARURL") == NULL) {
		strncpy(tmp, METARURL, URL_MAXSIZE);
	} else {
		strncpy(tmp, getenv("METARURL"), URL_MAXSIZE);
		if (verbose) printf("Using environment variable METARURL: %s\n", tmp);
	}

    if (snprintf(url, URL_MAXSIZE, "%s/%s.TXT", tmp, strupc(station)) < 0) 
        return 1;
	if (verbose) printf("Retrieving URL %s\n", url);

    curl_easy_setopt(curlhandle, CURLOPT_URL, url);
	curl_easy_setopt(curlhandle, CURLOPT_WRITEFUNCTION, receiveData);
	memset(noaabuffer, 0x0, METAR_MAXSIZE);

	res = curl_easy_perform(curlhandle);
	curl_easy_cleanup(curlhandle);

    return 0;
}


/* decode metar */
void decode_Metar(metar_t metar) {
	cloudlist_t *curcloud;
	obslist_t   *curobs; 
	int n = 0;
	double qnh;

	printf("Station       : %s\n", metar.station);
	printf("Day           : %i\n", metar.day);
	printf("Time          : %02i:%02i UTC\n", metar.time/100, metar.time%100);
	if (metar.winddir == -1) {
		printf("Wind direction: Variable\n");
	} else {
		static const char *winddirs[] = {
			"N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
			"S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"
	};
	n = ((metar.winddir * 4 + 45) / 90) % 16;
	printf("Wind direction: %i (%s)\n", metar.winddir, winddirs[n]);
	}
	printf("Wind speed    : %i %s\n", metar.windstr, metar.windunit);
	printf("Wind gust     : %i %s\n", metar.windgust, metar.windunit);
	printf("Visibility    : %i %s\n", metar.vis, metar.visunit);
	printf("Temperature   : %i C\n", metar.temp);
	printf("Dewpoint      : %i C\n", metar.dewp);

	qnh = metar.qnh;
	for (n = 0; n < metar.qnhfp; n++)
		qnh /= 10.0;
	printf("Pressure      : %.*f %s\n", metar.qnhfp, qnh, metar.qnhunit);
		
	printf("Clouds        : ");
	n = 0;
	for (curcloud = metar.clouds; curcloud != NULL; curcloud=curcloud->next) {
		if (n++ == 0) printf("%s at %d00 ft\n", 
				curcloud->cloud->type, curcloud->cloud->level);
		else printf("%15s %s at %d00 ft\n", 
				" ",curcloud->cloud->type, curcloud->cloud->level);
	}
	if (!n) printf("\n");

	printf("Phenomena     : ");
	n = 0;
	for (curobs = metar.obs; curobs != NULL; curobs=curobs->next) {
		if (n++ == 0) printf("%s\n", curobs->obs);
		else printf("%15s %s\n", " ",curobs->obs);
	}
	if (!n) printf("\n");
}


int main(int argc, char* argv[]) {
	int  res=0;
	metar_t metar;
	noaa_t  noaa;

	/* get options */
	opterr=0;
	if (argc == 1) {
		usage(argv[0]);
		return 1;
	}

	while ((res = getopt(argc, argv, "hvd")) != -1) {
		switch (res) {
			case '?':
				usage(argv[0]);
				return 1;
				break;
			case 'h':
				usage(argv[0]);
				return 0;
				break;
			case 'd':
				decode=1;
				break;
			case 'v':
				verbose=1;
				break;
		}
	}
    
    curl_global_init(CURL_GLOBAL_DEFAULT);

	// clear out metar and noaa
	memset(&metar, 0x0, sizeof(metar_t));
	memset(&noaa, 0x0, sizeof(noaa_t));

	while (optind < argc) {
		res = download_Metar(argv[optind++]);
        if (res == 0) {
			parse_NOAA_data(noaabuffer, &noaa);
			printf("%s", noaa.report);
			if (decode) {
				parse_Metar(noaa.report, &metar);
				decode_Metar(metar);
			}
		} else 
			printf("Error: %d\n", res);
	}
    
    return 0;
}

// EOF


src/metar.h
/* metar.c -- metar decoder
   $Id: metar.h,v 1.6 2006/04/05 20:30:28 kees-guest Exp $
   Copyright 2004,2005 Kees Leune <kees@leune.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* max size for a URL */
#define  URL_MAXSIZE 300

/* max size for a NOAA report */
#define  METAR_MAXSIZE 512

/* where to fetch reports */
#define  METARURL "http://weather.noaa.gov/pub/data/observations/metar/stations"

/* clouds */
typedef struct {
	char type[3];
	int  level;
} cloud_t;

/* linked list of clouds */
typedef struct cloudlist_el {
	cloud_t *cloud;
	struct cloudlist_el *next;
} cloudlist_t;

/* linked list of clouds */
typedef struct obslist_el {
	char *obs;
	struct obslist_el *next;
} obslist_t;

/* reports will be translated to this struct */
typedef struct {
	char station[10];
	int  day;
	int  time;
	int  winddir;  // winddir == -1 signifies variable winds
	int  windstr;
	int  windgust;
	char windunit[5];
	int  vis;
	char visunit[5];
	int  qnh;
	char qnhunit[5];
	int  qnhfp;	// fixed-point decimal places
	int  temp;
	int  dewp;
	cloudlist_t *clouds;
	obslist_t *obs;
} metar_t;

typedef struct {
	char date[36];
	char report[1024];
} noaa_t;

/* Parse the METAR contain in the report string. Place the parsed report in
 * the metar struct.
 */
void parse_Metar(char *report, metar_t *metar);

/* parse the NOAA report contained in the noaa_data buffer. Place a parsed
 * data in the metar struct. 
 */
void parse_NOAA_data(char *noaa_data, noaa_t *noaa);  


src/metar.c
/* metar.c -- metar decoder
   $Id: metar.c,v 1.6 2006/10/30 22:18:58 kees-guest Exp $
   Copyright 2004,2005 Kees Leune <kees@leune.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#include <regex.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "metar.h"

extern int verbose;

struct observation {
	const char *code;
	const char *description;
};

typedef struct observation observation_t;

struct observation observations[] = {
	{"MI", "Shallow "},
	{"BL", "Blowing "},
	{"BC", "Patches "},
	{"SH", "Showers "},
	{"PR", "Partials "},
	{"DR", "Drifting "},
	{"TS", "Thunderstorm "},
	{"FZ", "Freezing "},
	{"DZ", "Drizzle "},
	{"IC", "Ice Crystals "},
	{"UP", "Unknown "},
	{"RA", "Rain "},
	{"PL", "Ice Pellets "},
	{"SN", "Snow "},
	{"GR", "Hail "},
	{"SG", "Snow Grains "},
	{"GS", "Small hail/snow pellets "},
	{"BR", "Mist "},
	{"SA", "Sand "},
	{"FU", "Smoke "},
	{"HZ", "Haze "},
	{"FG", "Fog "},
	{"VA", "Volcanic Ash "},
	{"PY", "Spray "},
	{"DU", "Widespread Dust "},
	{"SQ", "Squall "},
	{"FC", "Funnel Cloud "},
	{"SS", "Sand storm "},
	{"DS", "Dust storm "},
	{"PO", "Well developed dust/sand swirls "},
	{"VC", "Vicinity "}
};


/* Add a cloud to a list of clouds */
static void add_cloud(cloudlist_t **head, cloud_t *cloud) {
	cloudlist_t *current;

	if (*head == NULL) {
		*head = malloc(sizeof(cloudlist_t));
		current = *head;
		current->cloud = cloud;
		current->next = NULL;
		return;
	}
	current = *head;
	while (current->next != NULL) 
		current = current->next;
	
	current->next = (cloudlist_t *)malloc(sizeof(cloudlist_t));
	current = current->next;
	current->cloud = cloud;
	current->next = NULL;
} // add_cloud


/* Add observation */
static void add_observation(obslist_t **head, char *obs) {
	obslist_t *current;

	if (*head == NULL) {
		*head = malloc(sizeof(obslist_t));
		current = *head;
		current->obs = obs;
		current->next = NULL;
		return;
	}
	current = *head;
	while (current->next != NULL) 
		current = current->next;
	
	current->next = (obslist_t *)malloc(sizeof(obslist_t));
	current = current->next;
	current->obs = obs;
	current->next = NULL;
} // add_observation


/* build the observations regexp patters*/
static void get_observations_pattern(char *pattern, int len) {
	int i=0, p=0;
	int size = sizeof(observations) / sizeof(observation_t);

	for (i=0; i < size; i++) {
		// add observation to string if we have enough space left
		p = strlen(pattern);
		if (p + strlen(observations[i].code) < len)
			strcpy(&pattern[p], observations[i].code);

		// add separator
		p = strlen(pattern);
		if (p < len) pattern[p++] = '|';
	}

	// remove last |
	pattern[strlen(pattern)-1] = 0;
} 


/* get the description of code pattern */
static char *decode_obs(char *pattern) {
	int i=0;
	int size = sizeof(observations) / sizeof(observation_t);

	for (i=0; i < size; i++) 
		if (strncmp(pattern, observations[i].code, 2) == 0)
			return (char*) observations[i].description;

	return NULL;
}


/* Analyse the token which is provided and, when possible, set the
 * corresponding value in the metar struct
 */
static void analyse_token(char *token, metar_t *metar) {
	regex_t preg;
	regmatch_t pmatch[5];
	int size;
	char tmp[99];
	char obspattern[255];
	char obsp[275];

	if (verbose) printf("Parsing token `%s'\n", token);

	// find station
	if (metar->station[0] == 0) {
		if (regcomp(&preg, "^([A-Z]+)$", REG_EXTENDED)) {
			perror("parseMetar");
			exit(errno);
		}
		if (!regexec(&preg, token, 5, pmatch, 0)) {
			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			memcpy(metar->station, token+pmatch[1].rm_so, 
					(size < 10 ? size : 10));
			if (verbose) printf("   Found station %s\n", metar->station);
			return;
		}
	}

	// find day/time
	if ((int)metar->day == 0) {
		if (regcomp(&preg, "^([0-9]{2})([0-9]{4})Z$", REG_EXTENDED)) {
			perror("parseMetar");
			exit(errno);
		}
		if (!regexec(&preg, token, 5, pmatch, 0)) {
			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[1].rm_so, (size < 99 ? size : 99));
			sscanf(tmp, "%d", (int*)&metar->day);

			size = pmatch[2].rm_eo - pmatch[2].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[2].rm_so, (size < 99 ? size : 99));
			sscanf(tmp, "%d", (int*)&metar->time);
			if (verbose) printf("   Found Day/Time %d/%d\n", 
					metar->day, metar->time);

			return;
		}
	} // daytime

	// find wind
	if ((int)metar->winddir == 0) {
		if (regcomp(&preg, "^(VRB|[0-9]{3})([0-9]{2})(G[0-9]+)?(KT)$", 
					REG_EXTENDED)) {
			perror("parseMetar");
			exit(errno);
		}
		if (!regexec(&preg, token, 5, pmatch, 0)) {
			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			memset(tmp, 0x0, 99);
			if (size) {
				memcpy(tmp, token+pmatch[1].rm_so, (size < 99 ? size : 99));
				if (strstr(tmp, "VRB") == NULL) { // winddir
					sscanf(tmp, "%d", (int*)&metar->winddir);
				} else { // vrb
					metar->winddir = -1;
				}
			}

			size = pmatch[2].rm_eo - pmatch[2].rm_so;
			memset(tmp, 0x0, 99);
			if (size) {
				memcpy(tmp, token+pmatch[2].rm_so, (size < 99 ? size : 99));
				sscanf(tmp, "%d", (int*)&metar->windstr);
			}

			size = pmatch[3].rm_eo - pmatch[3].rm_so;
			memset(tmp, 0x0, 99);
			if (size) {
				memcpy(tmp, token+pmatch[3].rm_so+1, (size < 99 ? size-1 : 99));
				sscanf(tmp, "%d", (int*)&metar->windgust);
			} else {
				metar->windgust = metar->windstr;
			}
			
			size = pmatch[4].rm_eo - pmatch[4].rm_so;
			if (size) {
				memcpy(&metar->windunit, token+pmatch[4].rm_so, 
					(size < 5 ? size : 5));
			}
			
			if (verbose) printf("   Found Winddir/str/gust/unit %d/%d/%d/%s\n", 
					metar->winddir, metar->windstr, metar->windgust,
					metar->windunit);
			return;
		}
	} // wind

	// find visibility
	if ((int)metar->vis == 0) {
		if (regcomp(&preg, "^([0-9]+)(SM)?$", REG_EXTENDED)) {
			perror("parsemetar");
			exit(errno);
		}
		if (!regexec(&preg, token, 5, pmatch, 0)) {
			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[1].rm_so, (size < 99 ? size : 99));
			sscanf(tmp, "%d", (int*)&metar->vis);
			
			size = pmatch[2].rm_eo - pmatch[2].rm_so;
			if (size) {
				memset(tmp, 0x0, 99);
				memcpy(&metar->visunit, token+pmatch[2].rm_so, 
						(size < 5 ? size : 5));
			} else 
				strncpy(metar->visunit, "M", 1);

			if (verbose) printf("   Visibility range/unit %d/%s\n", metar->vis, 
					metar->visunit);
			return;
		}
	} // visibility

	// find temperature and dewpoint
	if ((int)metar->temp == 0) {
		if (regcomp(&preg, "^(M?)([0-9]+)/(M?)([0-9]+)$", REG_EXTENDED)) {
			perror("parsemetar");
			exit(errno);
		}
		if (!regexec(&preg, token, 5, pmatch, 0)) {
			size = pmatch[2].rm_eo - pmatch[2].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[2].rm_so, (size < 99 ? size : 99));
			sscanf(tmp, "%d", &metar->temp);

			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[1].rm_so, (size < 99 ? size : 99));
			if (strncmp(tmp, "M", 1) == 0) metar->temp = -metar->temp;
			
			size = pmatch[4].rm_eo - pmatch[4].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[4].rm_so, (size < 99 ? size : 99));
			sscanf(tmp, "%d", &metar->dewp);
			
			size = pmatch[3].rm_eo - pmatch[3].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[3].rm_so, (size < 99 ? size : 99));
			if (strncmp(tmp, "M", 1) == 0) metar->dewp = -metar->dewp;

			if (verbose)
				printf("   Temp/dewpoint %d/%d\n", metar->temp, metar->dewp);
			return;
		}
	} // temp

	// find qnh
	if ((int)metar->qnh == 0) {
		if (regcomp(&preg, "^([QA])([0-9]+)$", REG_EXTENDED)) {
			perror("parsemetar");
			exit(errno);
		}
		if (!regexec(&preg, token, 5, pmatch, 0)) {
			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[1].rm_so, (size < 5 ? size : 5));
			if (strncmp(tmp, "Q", 1) == 0)
				strncpy(metar->qnhunit, "hPa", 3);
			else if (strncmp(tmp, "A", 1) == 0) {
				strncpy(metar->qnhunit, "\"Hg", 3);
				metar->qnhfp = 2;
			}
			else 
				strncpy(metar->qnhunit, "Unkn", 4);

			size = pmatch[2].rm_eo - pmatch[2].rm_so;
			memset(tmp, 0x0, 99);
			memcpy(tmp, token+pmatch[2].rm_so, (size < 99 ? size : 99));
			sscanf(tmp, "%d", &metar->qnh);

			if (verbose) 
				printf("   Pressure/unit %d/%s\n", metar->qnh, metar->qnhunit);

			return;
		}
	} // qnh

	// multiple cloud layers possible
	if (regcomp(&preg, "^(SKC|FEW|SCT|BKN|OVC)([0-9]{3})$", REG_EXTENDED)) {
		perror("parsemetar");
		exit(errno);
	}
	if (!regexec(&preg, token, 5, pmatch, 0)) {
		cloud_t *cloud = malloc(sizeof(cloud_t));
		memset(cloud, 0x0, sizeof(cloud_t));

		size=pmatch[1].rm_eo - pmatch[1].rm_so;
		memcpy(&cloud->type, token+pmatch[1].rm_so, (size < 3 ? size : 3));
		
		size=pmatch[2].rm_eo - pmatch[2].rm_so;
		memset(tmp, 0x0, 99);
		memcpy(tmp, token+pmatch[2].rm_so, (size < 3 ? size : 3));
		sscanf(tmp, "%d", (int*)&cloud->level);

		add_cloud((cloudlist_t **)&metar->clouds, cloud);
		if (verbose)
			printf("   Cloud cover/alt %s/%d00\n", cloud->type, cloud->level);
		return;
	} // cloud
	
	// phenomena
	memset(obsp, 0x0, 275);
	memset(obspattern, 0x0, 255);
	get_observations_pattern(obspattern, 255);
	snprintf(obsp, 255, "^([+-]?)((%s)+)$", obspattern);

	// cannot to CAVOK as an observation in the array because it is more than
	// 2 characters long and that screw up my algorithm
	if (strstr(token, "CAVOK") != NULL) {
		add_observation((obslist_t **)&metar->obs, "Ceiling and visibility OK");
	};

	if (regcomp(&preg, obsp, REG_EXTENDED)) {
		perror("parsemetar");
		exit(errno);
	}
	if (!regexec(&preg, token, 5, pmatch, 0)) {
		char *obs;
		obs = malloc(99);
		memset(obs, 0x0, 99);

		size=pmatch[1].rm_eo - pmatch[1].rm_so;
		memset(tmp, 0x0, 99);
		memcpy(tmp, token+pmatch[1].rm_so, (size < 1 ? size : 1));
		if (tmp[0] == '-') strncpy(obs, "Light ", 99);
		else if (tmp[0] == '+') strncpy(obs, "Heavy ", 99);

		// split up in groups of 2 chars and decode per group
		size=pmatch[2].rm_eo - pmatch[2].rm_so;
		memset(tmp, 0x0, 99);
		memcpy(tmp, token+pmatch[2].rm_so, (size < 99 ? size : 99));

		int i=0;
		char code[2];
		while (i < strlen(tmp)) {
			memset(code, 0x0, 2);
			memcpy(code, tmp+i, 2);
			strncat(obs, decode_obs(code), 99);
			i += 2;
		}
		
		// remove trailing space
		obs[strlen(obs)-1]=0;
		add_observation((obslist_t **)&metar->obs, obs);
		if (verbose) 
			printf("   Phenomena %s\n", obs);
		
		return;
	}

	if (verbose) printf("   Unmatched token = %s\n", token);
}


/* PUBLIC--
 * Parse the METAR contain in the report string. Place the parsed report in
 * the metar struct.
 */
void parse_Metar(char *report, metar_t *metar) {
	char *token;
	char *last;

	// clear resutls
	memset(metar, 0x0, sizeof(metar_t));

	// strip trailing newlines
	while ((last = strrchr(report, '\n')) != NULL)
		memset(last, 0, 1);

	token = strtok(report, " ");
	while (token != NULL) {
		analyse_token(token, metar);
		token = strtok(NULL, " ");
	}

} // parse_Metar


/* parse the NOAA report contained in the noaa_data buffer. Place a parsed
 * data in the metar struct. 
 */
void parse_NOAA_data(char *noaa_data, noaa_t *noaa) {
	regex_t preg;
	regmatch_t pmatch[10];
	int size;

	if (regcomp(&preg, "^([0-9/]+ [0-9:]+)[[:space:]]+(.*)$", REG_EXTENDED)) {
		fprintf(stderr, "Unable to compile regular expression.\n");
		exit(100);
	}

	if (regexec(&preg, noaa_data, 10, pmatch, 0)) {
		fprintf(stderr, "METAR pattern not found in NOAA data.\n");
	} else {
		memset(noaa, 0x0, sizeof(noaa_t));
		/* date */
        size = pmatch[1].rm_eo - pmatch[1].rm_so;
        memcpy(noaa->date, noaa_data+pmatch[1].rm_so, (size < 36 ? size : 36));

		/* metar */
		size = pmatch[2].rm_eo - pmatch[2].rm_so;
		memcpy(noaa->report, noaa_data+pmatch[2].rm_so, 
				(size < 1024 ? size : 1024));
	}
} // parse_NOAA_data
